forgevtt-module.js

```diff
@@ -1872,80 +1879,96 @@ class ForgeVTT_FilePicker extends FilePicker {
             // from the assets library, even if it's a relative path
             const dataDirs = ["systems", "modules"];
             const publicDirs = ["cards", "icons", "sounds", "ui"];
             if ([...dataDirs, ...publicDirs].every((folder) => !target.startsWith(`${folder}/`))) {
-                return ["forgevtt", target];
+                return ["forgevtt", target, undefined];
             }
         }
-        return super._inferCurrentDirectory(target);
+        return super._inferCurrentDirectory(_target);
     }
-    /* eslint-enable no-param-reassign */
 
+    /**
+     * Determines whether the user can upload assets.
+     * @returns {boolean} Returns true if the user can upload assets, otherwise false.
+     */
     get canUpload() {
         if (this.activeSource === "forgevtt") {
-            if (this.source.bucket === "my-assets") {
-                return true;
-            }
-            const bucket = this._forgeBucketIndex.find((b) => b.key === this.source.bucket);
-            const apiKey = bucket ? bucket.jwt : null;
-            if (!ForgeAPI.isValidAPIKey(apiKey)) {
-                return false;
+            const bucket = this.constructor._getForgeVttBucket(this.source.bucket);
+            if (bucket) {
+                if (bucket.key === "my-assets") {
+                    return true;
+                }
+                if (!ForgeAPI.isValidAPIKey(bucket.jwt)) {
+                    return false;
+                }
+                const permissions = ForgeAPI._tokenToInfo(bucket.jwt).permissions || [];
+                return permissions.includes("write-assets");
             }
-            const info = ForgeAPI._tokenToInfo(apiKey);
-            return (info.permissions || []).includes("write-assets");
         }
         if (this.activeSource === "forge-bazaar") {
             return false;
         }
         return !ForgeVTT.usingTheForge && super.canUpload;
     }
 
-    /* Override _onChangeBucket to use the current source instead of the hardcoded s3 source
+    /**
+     * Intercept and stop Foundry's #onChangeBucket "change" event to use our own handler instead
+     * @param {Event} event - The change event object.
+     * @returns {Promise<void>} - A promise that resolves when the browsing is complete.
      */
-    _onChangeBucket(event) {
+    async _onChangeBucketParent(event) {
         event.preventDefault();
-        const select = event.currentTarget;
-        this.sources[this.activeSource].bucket = select.value;
-        return this.browse("/");
+        event.stopImmediatePropagation();
+        event.stopPropagation();
+        if (event.target.name === "bucket") {
+            const select = event.target;
+            select.disabled = true;
+            this.activeSource = "forgevtt";
+            this.source.bucket = select.value;
+            this.sources.forgevtt.bucket = select.value;
+            await this.browse("/");
+            select.disabled = false;
+        }
     }
```

Comment:
The intention here is to stop Foundry's `#onChangeBucket`, but it is still triggered somehow.

Reply:
The name changed to `_onChangeBucketParent`, so it is probably no longer overriding the original `onChangeBucket`.

Reply:
Foundry's `_onChangeBucket` was renamed to `#onChangeBucket`, making it a private method and if I understand correctly the "overriding" in place doesn't work anymore. And to maintain compatibility with <v12 I thought it could be smart to intercept it from the parent element but obviously the event bubbles up from child to parent so that doesn't really work as intended.

Reply:
Instead of "overriding" it, Foundry's event is now cancelled and reset to our own handler.